### ISSUE TITLE
transformations inside of init

### ISSUE THREAD

**Issue by Howuhh** [patrick-kidger/equinox#377]
Hi! At one point, I thought the code would be clearer if I applied the right transformations to the layers immediately instead of on the forward pass. For example, my encoder which expect 2D batch (just an example):
```python
class Encoder(eqx.Module):
    encoder: nn.Linear

    def __init__(self, input_dim, *, key):
        self.encoder = eqx.filter_vmap(nn.Linear(input_dim, 256, key=key))

    def __call__(self, obs):
        # obs: [batch_size, input_dim]
        emb = self.encoder(obs)
        return emb
```
Although the normal `apply_updates` works (at least it doesn't throw errors), in RL it is often necessary to update weights in this way:
```python
@eqx.filter_jit
def example_soft_update(model):
    params, static = eqx.partition(model, eqx.is_array)
    new_params = optax.incremental_update(params, params, 0.1)   # usually there will be (params, target_params), not same weights
    return eqx.combine(new_params, static)
```
And with module with vmaped layer inside this will raise an error. Should I filter params more rigorously or this is just a bad approach in general?

**Comment by patrick-kidger** [comment]
I'm assuming your first example is meant to use `self.encoder = eqx.filter_vmap(nn.Linear)(input_dim, 256, key=key)`? In which case, are you really wanting to use a different encoder for each batch element? That'd be very unusual.

In your second code snippet, you probably meant to write `params, static = eqx.partition(...)` rather than `eqx.filter(...)`. That's probably the source of the error.

There's nothing wrong with using `filter_vmap`'d modules! You just need to be careful where you use them.

**Comment by Howuhh** [comment]
No! Actually I want same encoder to each element in the batch (I was a little too lazy to write in the call vmap every time on the batches and that's how I stumbled upon this error)

**Comment by Howuhh** [comment]
This is standalone example which I use:
```python
import jax
import equinox as eqx
import equinox.nn as nn
import optax

class TestModel(eqx.Module):
    encoder: nn.Linear

    def __init__(self, in_channels, *, key):
        keys = jax.random.split(key, 3)
        self.encoder = eqx.filter_vmap(nn.Linear(in_channels, 265, key=keys[0]))

    def __call__(self, obs):
        emb = self.encoder(obs)
        return emb
        
@eqx.filter_jit
def update(critic, target_critic):
    params, static = eqx.partition(critic, eqx.is_array)
    target_params, target_static = eqx.partition(target_critic, eqx.is_array)

    new_target_params = optax.incremental_update(params, target_params, 0.1)
    new_target_criitc = eqx.combine(new_target_params, target_static)

    return new_target_criitc
        
        
key = jax.random.PRNGKey(0)
critic = TestModel(3, key=key)
target_critic = TestModel(3, key=key)

# this will raise an error
new_critic = update(critic, target_critic)
```

**Comment by Howuhh** [comment]
I think I got it....
```
eqx.filter_vmap(nn.Linear)(3, 256, key=key)  # this is same linear vmaped on full batch

eqx.filter_vmap(nn.Linear(3, 256, key=key)) # this is different linear for each example in the batch
```
However, I cannot use first variant as `keyword arguments cannot be used with functions wrapped with filter_vmap` for the key=key and key always should be keyword argument...

Well, I guess we should always do vmap in __call__

**Comment by patrick-kidger** [comment]
> ```
> eqx.filter_vmap(nn.Linear)(3, 256, key=key)  # this is same linear vmaped on full batch
>
> eqx.filter_vmap(nn.Linear(3, 256, key=key)) # this is different linear for each example in the batch
> ```

Other way around! The first one creates a batch of `Linear` layers; the second one batches the single `Linear` layer that is created.

> However, I cannot use first variant as `keyword arguments cannot be used with functions wrapped with filter_vmap` for the key=key and key always should be keyword argument...

Ah, that's annoying. That can be worked around by doing  `eqx.filter_vmap(lambda *args, key: nn.Linear(*args, key=key))(3, 256, key)` instead, but I can see that that's quite annoying. Maybe we should do the same thing JAX does and just unconditionally set `in_axes=0` for all keyword arguments.

---

> This is standalone example which I use:

Thanks for the repro! I've ended up tracking this down to a incompatbility in the three-way combination of `dataclass` and `property` and `__wrapped__`. (All Python built-ins.) I'll have a think about what a possible fix might be.

**Comment by patrick-kidger** [comment]
Alright, this should be fixed in #381!
